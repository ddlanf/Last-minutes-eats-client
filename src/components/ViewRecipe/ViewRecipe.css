.view-recipe-recipe{
    margin-top: 18%;
    margin-right: auto;
    margin-left: auto;
    width: 90%;
    max-width: 2000px;
}

.view-recipe-image-box{
    margin-top: 5%;
    margin-left: auto;
    margin-right: auto;
    width: 90%;
    text-align: center;
}

.view-recipe-image{
    width: 100%;
    margin-right: auto;
    margin-left: auto;
}
.view-recipe-main-detail > *{
    margin-left: auto;
    margin-right: auto;
    text-align: center;
    margin-top: 2%;
    font-size: 5vw;
    margin-bottom: 1%;
}

.view-recipe-recipe-name-desktop{
    display: none;
}

.view-recipe-ingredients, .view-recipe-instructions{
    text-align: left;
    margin: 0;
    padding-left: 14%;
    width: 90%;
    font-size: 4.5vw;
}

.view-recipe-rating-desktop{
    display: none;
}

.view-recipe-back-and-edit{
    font-size: 5.5vw;
    width: 90%;
    margin-left: auto;
    margin-right: auto;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-content: center;
}

a.view-recipe-edit, a.view-recipe-back{
    color: black;
    text-decoration: none;
    margin-top: 6%;
    text-align: center;
    margin-bottom: 3%;
}

.view-recipe-delete{
    display: block;
    background-color: white;
    border: white solid 1px;
    color: red;
    text-decoration: none;
    font-size: 3.5vw;
    margin-right: auto;
    margin-left: auto;
    margin-bottom: 5%;
    text-align: center;
}

.view-recipe-preparation-time-mobile{
    display: block;
    font-size: 4.5vw;
}

.view-recipe-preparation-time-desktop{
    display: none;
    font-size: 4.5vw;
}

.view-recipe-star{
    color: gold;
}
.view-recipe-star-box-mobile{
    text-align: center;
    margin-top: 5%;
    width: 100%;
}

.view-recipe-star-box-desktop{
    display: none;
}


@media screen and (min-width: 1000px) {

    .view-recipe-image{
        max-height: 500px;
    }
   
    .view-recipe-recipe{
        margin-top: 10%;
        display: flex;
        flex-direction: row-reverse;
        flex-wrap: wrap;
    }

    .view-recipe-rating-desktop{
        display: block;
        font-size: 3.5vw;
    }

    .view-recipe-rating-mobile{
        display: none;
    }

    .view-recipe-main{
        width: 50%;
        text-align: center;
    }

    .view-recipe-recipe-name-mobile{
        display: none;
    }

    .view-recipe-recipe-name-desktop{
        display: block;
        font-size: 3.5vw;
    }

    h2.view-recipe-ingredients-header, h2.view-recipe-instructions-header{
        font-size: 3.2vw;
    }

    .view-recipe-ingredients.short-list{
        font-size: 2.8vw;
    }
    .view-recipe-ingredients.long-list{
        font-size: 2.2vw;
    }

    .view-recipe-instructions.short-list{
        font-size: 2.3vw;
    }

    .view-recipe-instructions.long-list{
        font-size: 2vw;
    }

    .view-recipe-main-detail{
        width: 50%;
    }

    .view-recipe-main-detail > *{
        font-size: 4vw;
    }

    .view-recipe-recipe-name-desktop{
        margin-top: 4%;
        margin-bottom: 0;
    }

    .view-recipe-preparation-time-mobile{
        display: none;
    }
    
    .view-recipe-preparation-time-desktop{
        display: block;
        font-size: 2.5vw;
        margin-top: 2%;
    }

    .view-recipe-image-box{
        max-height: 550px;
        max-width: 550px;
    }

    a.view-recipe-edit, a.view-recipe-back{
        font-size: 3vw;
    }

    
    .view-recipe-star-box-mobile{
        display: none;
    }
    
    .view-recipe-star-box-desktop{
        display: block;
        margin-top: 0;
    }  

    a.view-recipe-delete{
        font-size: 2.2vw;
    }
}
