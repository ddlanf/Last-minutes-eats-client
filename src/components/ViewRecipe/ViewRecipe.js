import React, { Component } from 'react'
import { Link, withRouter } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import RecipesApiService from '../../services/recipes-api-service'
import RecipeContext from '../../contexts/RecipeContext'
import StarRating from './StarRating/StarRating'
import DeleteRecipe from './DeleteRecipe/DeleteRecipe'
import './ViewRecipe.css'

class ViewRecipe extends Component {

    static contextType = RecipeContext

    state = { showDeleteRecipe: false }

    makeIngredientsList(ingredients){
        return ingredients.map((ingredient, index) =>{
            return <li key={index}>{ingredient}</li>
        })
    }

    makeStepsList(steps){
        return steps.map((step, index)=>{
            return <li key={index}>{step}</li>
        })
    }


    makeStars(numberOfStars){
        const stars = []
        
        for(let i = 1; i < 6; i++){

            if(numberOfStars < i  && numberOfStars < i - 0.75){
                stars.push((<FontAwesomeIcon key={i} className="view-recipe-star" icon={['far', 'star']}/>))
            }
            else if(numberOfStars >= i - 0.75 && numberOfStars <= i - 0.5){
                stars.push((<FontAwesomeIcon key={i} className="view-recipe-star" icon="star-half-alt"/>))
            }
            else if(numberOfStars > i - 0.5 && numberOfStars < i - 0.25){
                stars.push((<FontAwesomeIcon key={i} className="view-recipe-star" icon="star-half-alt"/>))
            }
            else{  stars.push((<FontAwesomeIcon  key={i} className="view-recipe-star" icon={['fas', 'star']}/>))  }
        }

        return stars
    }

    toggleDeleteRecipe = (showDeleteRecipe) =>{
        this.setState({ showDeleteRecipe })
    }

    componentDidMount(){
        const { recipeId } = this.props.match.params;

        RecipesApiService.getRecipe(recipeId)
            .then(recipe =>{
                this.context.setRecipe(recipe)
            })
            .catch(res => {this.context.setError(res.error)})
    }

    render() {
        
        const defaultRecipe = { steps: [], ingredients: [], preparation_time_unit: ''}

        const recipe = this.context.recipe.recipe_name ? this.context.recipe : defaultRecipe 
     
        const { recipeId } = this.props.match.params;

        return (
            <>
                <DeleteRecipe
                    showDeleteRecipe={this.state.showDeleteRecipe}
                    toggleDeleteRecipe={this.toggleDeleteRecipe}
                />
                <section className="view-recipe-recipe">
                    <div className="view-recipe-main">
                        <h1 className="view-recipe-recipe-name-desktop">{recipe.recipe_name}</h1>
                        <div className="view-recipe-rating-desktop">
                          {this.makeStars(recipe.overall_rating)}
                        </div>
                        <div className="view-recipe-image-box"> 
                            <img className="view-recipe-image" alt={recipe.recipe_name} src={recipe.image} />
                        </div>
                        <h2 className="view-recipe-preparation-time-desktop">
                            {recipe.preparation_time + recipe.preparation_time_unit.slice(0, 3)}
                        </h2>
                        <div className="view-recipe-star-box-desktop">
                            <StarRating recipeId={recipeId}/>
                        </div>
                    </div>
                    
                    <div className="view-recipe-main-detail">
                        <h1 className="view-recipe-recipe-name-mobile">{recipe.recipe_name}</h1>
                        <div className="view-recipe-rating-mobile">
                            {this.makeStars(recipe.overall_rating)}
                        </div>
                        <h2 className="view-recipe-preparation-time-mobile">
                            {recipe.preparation_time} {recipe.preparation_time_unit.slice(0, 3)}
                        </h2>
                        <h2 className="view-recipe-ingredients-header">Ingredients</h2>
                        <ul 
                            className={`view-recipe-ingredients  ${(recipe.ingredients.length <= 3) ? 'short-list' : 'long-list'}`}>
                            {this.makeIngredientsList(recipe.ingredients)}
                        </ul>
                        <h2 className="view-recipe-ingredients-header">Instructions</h2>
                        <ol className={`view-recipe-instructions
                         ${(recipe.steps.length <= 4) ? 'short-list' : 'long-list'}`}>
                            {this.makeStepsList(recipe.steps)}
                        </ol>
                    </div>
                    <div className="view-recipe-star-box-mobile">
                        <StarRating recipeId={recipeId}/>
                    </div>
                    <div className="view-recipe-back-and-edit">
                        <Link
                            to="/view-all-recipes"
                            className="view-recipe-back"
                            >
                            Back
                        </Link>
                        <Link 
                            to={`/edit-recipe/${recipe.id}`}
                            className="view-recipe-edit">
                            Edit
                        </Link>
                    </div>
                </section>  
                <button
                    onClick={() => this.toggleDeleteRecipe(true)}
                    className="view-recipe-delete"
                    >
                    Delete this recipe
                </button>
            </>
        )
    }
}

export default withRouter(ViewRecipe)