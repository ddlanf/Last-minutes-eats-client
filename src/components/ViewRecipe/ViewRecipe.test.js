import React from 'react';
import ReactDOM from 'react-dom';
import ViewRecipe from './ViewRecipe';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom'
 
export function FontAwesomeIcon(props) {
    return <i className="fa"/>
}

describe('Make component', () => {


    it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                        <ViewRecipe/>
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const tree = renderer
          .create(
                <BrowserRouter>
                    <ViewRecipe/>
                </BrowserRouter>)
          .toJSON();
        expect(tree).toMatchSnapshot();  
      });
});