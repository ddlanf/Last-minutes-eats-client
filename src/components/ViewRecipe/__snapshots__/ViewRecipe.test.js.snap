// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make component renders the UI as expected 1`] = `
Array [
  <section
    className="view-recipe-logo"
  >
    <div
      className="view-recipe-food-image"
    />
  </section>,
  <section
    className="view-recipe-recipe"
  >
    <div
      className="view-recipe-main"
    >
      <h1
        className="view-recipe-recipe-name-desktop"
      />
      <div
        className="view-recipe-rating-desktop"
      >
        <i
          className="fa"
        />
        <i
          className="fa"
        />
        <i
          className="fa"
        />
        <i
          className="fa"
        />
        <i
          className="fa"
        />
      </div>
      <div
        className="view-recipe-image-box"
      >
        <img
          className="view-recipe-image"
        />
      </div>
      <h2
        className="view-recipe-preparation-time-desktop"
      >
        undefined
      </h2>
      <div
        className="view-recipe-star-box-desktop"
      >
        <label
          className="view-recipe-rate"
        >
          Your rating
        </label>
        <span
          className="view-recipe-stars"
        >
          <div
            className="star-rating"
          >
            <i
              className="fa"
            />
            <i
              className="fa"
            />
          </div>
          <div
            className="star-rating"
          >
            <i
              className="fa"
            />
            <i
              className="fa"
            />
          </div>
          <div
            className="star-rating"
          >
            <i
              className="fa"
            />
            <i
              className="fa"
            />
          </div>
          <div
            className="star-rating"
          >
            <i
              className="fa"
            />
            <i
              className="fa"
            />
          </div>
          <div
            className="star-rating"
          >
            <i
              className="fa"
            />
            <i
              className="fa"
            />
          </div>
        </span>
      </div>
    </div>
    <div
      className="view-recipe-main-detail"
    >
      <h1
        className="view-recipe-recipe-name-mobile"
      />
      <div
        className="view-recipe-rating-mobile"
      >
        <i
          className="fa"
        />
        <i
          className="fa"
        />
        <i
          className="fa"
        />
        <i
          className="fa"
        />
        <i
          className="fa"
        />
      </div>
      <h2
        className="view-recipe-preparation-time-mobile"
      >
        undefined
      </h2>
      <h2
        className="view-recipe-ingredients-header"
      >
        Ingredients
      </h2>
      <ul
        className="view-recipe-ingredients"
      />
      <h2>
        Instructions
      </h2>
      <ol
        className="view-recipe-instructions"
      />
    </div>
    <div
      className="view-recipe-star-box-mobile"
    >
      <label
        className="view-recipe-rate"
      >
        Your rating
      </label>
      <span
        className="view-recipe-stars"
      >
        <div
          className="star-rating"
        >
          <i
            className="fa"
          />
          <i
            className="fa"
          />
        </div>
        <div
          className="star-rating"
        >
          <i
            className="fa"
          />
          <i
            className="fa"
          />
        </div>
        <div
          className="star-rating"
        >
          <i
            className="fa"
          />
          <i
            className="fa"
          />
        </div>
        <div
          className="star-rating"
        >
          <i
            className="fa"
          />
          <i
            className="fa"
          />
        </div>
        <div
          className="star-rating"
        >
          <i
            className="fa"
          />
          <i
            className="fa"
          />
        </div>
      </span>
    </div>
    <div
      className="view-recipe-back-and-edit"
    >
      <a
        className="view-recipe-back"
        href="/view-all-recipes"
        onClick={[Function]}
      >
        Back
      </a>
      <a
        className="view-recipe-edit"
        href="/edit-recipe/undefined"
        onClick={[Function]}
      >
        Edit
      </a>
    </div>
  </section>,
]
`;
